File: src/main/java/gargoyle/ct/resource/internal/LocalResource.java
package gargoyle.ct.resource.internal;

import gargoyle.ct.resource.Resource;
import gargoyle.ct.util.ResourceException;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.io.File;
import java.io.IOException;
import java.io.Serializable;
import java.net.MalformedURLException;
import java.net.URL;
import java.text.MessageFormat;
import java.util.Comparator;

/**
 * local filesystem {@link Resource} from different locations
 */
public class LocalResource extends VirtualResource {
    private static final String ENV_USER_DIR = "user.dir";
    private static final String ENV_USER_HOME = "user.home";
    private static final String MSG_CANNOT_CREATE_ROOTS = "Cannot create roots";
    private static final String MSG_CANNOT_USE_0_AS_ROOT = "Cannot use {0} as root";

    protected LocalResource(URL url) {
        super(url.toExternalForm());
    }

    @Nullable
    @SuppressWarnings("ObjectAllocationInLoop")
    protected static LocalResource findLocal(@NotNull String name) {
        LocalResource writable = null;
        int writablePriority = Integer.MIN_VALUE;
        for (LocalLocation root : getReadableLocations()) {
            if (root != null && root.getUrl() != null) {
                try {
                    LocalResource resource = new LocalResource(new URL(root.getUrl(), name));
                    if (resource.exists()) {
                        return resource;
                    }
                    int writePriority = root.getWritePriority();
                    if (resource.isWritable() && (writable == null || writablePriority < writePriority)) {
                        writable = resource;
                        writablePriority = writePriority;
                    }
                } catch (MalformedURLException ex) {
                    throw new ResourceException(MessageFormat.format(MSG_CANNOT_USE_0_AS_ROOT, root), ex);
                }
            }
        }
        return writable;
    }

    private static LocalLocation[] getReadableLocations() {
        return getLocations();
    }

    private static LocalLocation[] getLocations() {
        try {
            return new LocalLocation[]{
                    new LocalLocation(new File(".").toURI().toURL(), 0, 0),
                    new LocalLocation(new File(System.getProperty(ENV_USER_DIR, DOT)).toURI().toURL(), 0, 0),
                    new LocalLocation(getHomeDirectoryLocation(), 0, 1),
                    new LocalLocation(LocalResource.class.getResource("/"), 0, 0)
            };
        } catch (MalformedURLException ex) {
            throw new ResourceException(MSG_CANNOT_CREATE_ROOTS, ex);
        }
    }

    private static LocalLocation[] getWritableLocations() {
        return getLocations();
    }

    public static URL getHomeDirectoryLocation() {
        try {
            return new File(System.getProperty(ENV_USER_HOME, DOT)).toURI().toURL();
        } catch (MalformedURLException e) {
            throw new ResourceException(e.getLocalizedMessage(), e);
        }
    }

    @Nullable
    public File toFile() throws IOException {
        URL url = toURL();
        if (url == null) {
            return null;
        }
        String path = url.getPath();
        return new File(path.startsWith(File.separator) ? path.substring(1) : path); // XXX hacky
    }

    private static class LocalLocation {
        private final int readPriority;
        private final URL url;
        private final int writePriority;

        public LocalLocation(URL url, int readPriority, int writePriority) {
            this.url = url;
            this.readPriority = readPriority;
            this.writePriority = writePriority;
        }

        public int getReadPriority() {
            return readPriority;
        }

        public URL getUrl() {
            return url;
        }

        public int getWritePriority() {
            return writePriority;
        }

        @Override
        public int hashCode() {
            return url.toExternalForm().hashCode();
        }

        @Override
        public boolean equals(@Nullable Object obj) {
            if (this == obj) return true;
            if (obj == null || getClass() != obj.getClass()) return false;
            LocalLocation that = (LocalLocation) obj;
            return url.toExternalForm().equals(that.url.toExternalForm());
        }
    }

    private static class LocalLocationReadPriority implements Comparator<LocalLocation>, Serializable {
        private static final long serialVersionUID = 3293380242047042026L;

        @Override
        public int compare(@NotNull LocalLocation o1, @NotNull LocalLocation o2) {
            return Integer.compare(o1.getReadPriority(), o2.getReadPriority());
        }
    }

    private static class LocalLocationWritePriority implements Comparator<LocalLocation>, Serializable {
        private static final long serialVersionUID = 1771423086301596535L;

        @Override
        public int compare(@NotNull LocalLocation o1, @NotNull LocalLocation o2) {
            return Integer.compare(o1.getWritePriority(), o2.getWritePriority());
        }
    }
}

File: util/src/main/kotlin/gargoyle/ct/util/resource/internal/LocalResource.kt
package gargoyle.ct.util.resource.internal

import gargoyle.ct.util.util.ResourceException
import java.io.File
import java.net.MalformedURLException
import java.net.URL
import java.text.MessageFormat

/**
 * local filesystem [Resource] from different locations
 */
open class LocalResource protected constructor(url: URL) : VirtualResource<LocalResource>(url.toExternalForm()) {
    fun toFile(): File {
        val url = toURL()
        val path = url.path
        return File(if (path.startsWith(File.separator)) path.substring(1) else path) // XXX hacky
    }

    private class LocalLocation(private val url: URL, val readPriority: Int, val writePriority: Int) {

        fun getUrl(): URL {
            return url
        }

        override fun hashCode(): Int {
            return url.toExternalForm().hashCode()
        }

        override fun equals(other: Any?): Boolean =
            when {
                this === other -> true
                other == null || javaClass != other.javaClass -> false
                else -> {
                    val that = other as LocalLocation
                    url.toExternalForm() == that.url.toExternalForm()
                }
            }
    }

    private class LocalLocationReadPriority : Comparator<LocalLocation> {
        override fun compare(o1: LocalLocation, o2: LocalLocation): Int {
            return o1.readPriority.compareTo(o2.readPriority)
        }
    }

    private class LocalLocationWritePriority : Comparator<LocalLocation> {
        override fun compare(o1: LocalLocation, o2: LocalLocation): Int = o1.writePriority.compareTo(o2.writePriority)
    }

    companion object {
        private const val ENV_USER_DIR = "user.dir"
        private const val ENV_USER_HOME = "user.home"
        private const val MSG_CANNOT_CREATE_ROOTS = "Cannot create roots"
        private const val MSG_CANNOT_USE_0_AS_ROOT = "Cannot use {0} as root"
        fun findLocal(name: String): LocalResource {
            var writable: LocalResource? = null
            var writablePriority = Int.MIN_VALUE
            for (root in readableLocations) {
                try {
                    val resource = LocalResource(URL(root.getUrl(), name))
                    if (resource.exists()) {
                        return resource
                    }
                    val writePriority: Int = root.writePriority
                    if (resource.isWritable && (writable == null || writablePriority < writePriority)) {
                        writable = resource
                        writablePriority = writePriority
                    }
                } catch (ex: MalformedURLException) {
                    throw ResourceException(MessageFormat.format(MSG_CANNOT_USE_0_AS_ROOT, root), ex)
                }
            }
            return writable!!
        }

        private val readableLocations: Array<LocalLocation>
            get() = locations
        private val locations: Array<LocalLocation>
            get() = try {
                arrayOf(
                    LocalLocation(File(".").toURI().toURL(), 0, 0),
                    LocalLocation(File(System.getProperty(ENV_USER_DIR, ".")).toURI().toURL(), 0, 0),
                    LocalLocation(homeDirectoryLocation, 0, 1),
                    //FIXME
//                    LocalLocation(
//                        URL(
//                            LocalLocation::class.getResource("/").toExternalForm().substringBeforeLast('/')
//                        ), 0, 0
//                    )
                )
            } catch (ex: MalformedURLException) {
                throw ResourceException(MSG_CANNOT_CREATE_ROOTS, ex)
            }
        private val writableLocations: Array<LocalLocation>
            get() = locations
        val homeDirectoryLocation: URL
            get() = try {
                File(System.getProperty(ENV_USER_HOME, ".")).toURI().toURL()
            } catch (e: MalformedURLException) {
                throw ResourceException(e.localizedMessage, e)
            }
    }
}
